riscv: add a renode-simulated virtual riscv soc

ycsin opened the pull request:
Proposing to add a Renode-simulated RISC-V 32BIT board which peripherals can be extended easily using Renode's text-based platform description script, to test some hardware configurations that only existed in some custom board / requires a recompilation of QEMU, for example a multi-instance PLIC RISC-V board, see #65288. [Board documentation preview](https://builds.zephyrproject.io/zephyr/pr/65500/docs/boards/riscv/renode_riscv32/doc/index.html)

> [!NOTE]
> Has commit from #66362 & #66419

- [x] Commits are abit messed up, will recommit them if we were to move it forward

carlocaione commented on boards/riscv/renode_riscv32/doc/index.rst:
```diff
@@ -0,0 +1,55 @@
+Flashing
+========
+
+While this board is emulated and you can't "flash" it, you can use this
+configuration to run basic Zephyr applications and kernel tests in the Renode
+emulated environment. For example, with the :zephyr:code-sample:`synchronization` sample:
```
please, add a paragraph about __how__ you are using renode to run this platform

ycsin replied:
I don't get it - it is similar to how `qemu_riscv32` works:
```
west build -b renode_riscv32 -p auto -t run samples/hello_world
```

https://docs.zephyrproject.org/latest/boards/riscv/qemu_riscv32/doc/index.html

carlocaione replied:
ok, I didn't realise we had a `renode_riscv32` board.

ycsin replied:
no, this PR is adding a new `renode_riscv32` board, it's just that the usage is similar to its `qemu_riscv32` equivalent

fkokosinski commented on boards/riscv/renode_riscv32/doc/index.rst:
```diff
@@ -0,0 +1,55 @@
+.. _renode-riscv32:
+
+Renode RISCV32
```
Hi! While Renode normally targets real-life hardware (i.e. same software is run on hw and in sim), it also allows its users to create virtual platforms, or to model hardware which is yet to be released.

I would suggest you to change the name of this virtual platform to something less generic sounding, so that it's less confusing for the user. I'm worried that it might be confusing for the user to see e.g. commands such as `west build -b hifive1 -t run_renode` and then see that there's a separate platform called `renode_riscv32` and cause them to wonder - how are these two things related?

If you have a name for this hw (any identification string, really), I'd suggest you to change the name of this Zephyr target to something like `<name>_virtual`. I see in the YAML file that you're referencing `m2gl025_miv` there. So maybe something like `m2gl025_miv_virtual`, and adding a description saying that this platform is derived from the `m2gl025_miv` Zephyr target would be a good idea?

ycsin replied:
My original intention was to create a board for [riscv_virt](https://github.com/renode/renode/blob/master/platforms/cpus/riscv_virt.repl) rather than `m2gl025_miv`, wouldn't naming it as `m2gl025_miv_virtual` cause confusion to the user of `m2gl025_miv`?

fkokosinski replied:
Ah, okay then, because `boards/riscv/renode_riscv32/renode_riscv32.yaml` looked really familiar to its counterpart for the `m2gl025_miv` board (e.g. `vendor` set to Microchip) to me. Maybe renaming the board to `riscv_virtual`, or `riscv_custom` would be fine with you?

ycsin replied:
> Maybe renaming the board to `riscv_virtual`, or `riscv_custom` would be fine with you?

`riscv_virtual` sounds good to me, how about the `soc`? Is the current name ok for you?

ycsin replied:
I've changed the `soc` from `renode` to `renode_virt`

cfriedt commented:
Weird how the shell rests fail

ycsin commented:
> Weird how the shell rests fail

That was due to wrong UART IRQ ↔ PLIC configuration in the Renode .repl file, should be fixed now

tgorochowik commented:
Could you rebase to fix the conflicts? Otherwise I think it looks good

ycsin committed:
dts: bindings: vendor-prefixes: add renode prefix

Add renode binding prefix.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
dts: riscv: add a SoC dtsi for Renode RISC-V Virt SoC

RISCV32 simulation dtsi for Renode.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
soc: riscv: add support for Renode Virt RISCV32 SoC

Add a beef-ed up version of Renode's `riscv_virt` SoC

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
boards: riscv: add RISCV32 Virtual board made with Renode

The RISCV32 Virtual board is a virtual platform made with
Renode as an alternative to QEMU. Contrary to QEMU, the
peripherals of this platform can be easily configured by
editing the `riscv32_virtual.repl` script and the devicetree
files accordingly, this allows certain hardware configurations
that only exist in proprietary boards/SoCs to be tested in
upstream CI.

Added another entry for this board to the excluded platform in
`kernel.timer.timer` test.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
> Could you rebase to fix the conflicts? Otherwise I think it looks good

rebased, PTAL
